### PR TITLE
chore: remove node from VMSS VM cache when it is deleted

### PR DESCRIPTION
**Pull request by nilo19**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Remove the node from the VMSS VM cache if it is deleted. This will help reduce the chance of conflict between aks and cloud provider azure.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/provider/azure_vmss_cache.go [hunk 1]
```diff
@@ -266,29 +266,52 @@ func (ss *ScaleSet) newVMSSVirtualMachinesCache(resourceGroupName, vmssName, cac
...
+	}
+	if managedByAS {
+		cached, err := ss.availabilitySetNodesCache.Get(consts.AvailabilitySetNodesKey, azcache.CacheReadTypeDefault)
+		if err != nil {
+			klog.Warningf("DeleteCacheForNode(%s): failed to get availabilitySetNodesCache: %s", err.Error())
+			return nil
+		}
+		entry := cached.(*availabilitySetNodeEntry)
+		entry.nodeNames.Delete(nodeName)
+		entry.vmNames.Delete(nodeName)
```
Comment (feiskyer) [inline pkg/provider/azure_vmss_cache.go:283]: do we need to invoke `az.vmCache.Delete(vmName)` here?

File: pkg/provider/azure_vmss_cache.go [hunk 2]
```diff
@@ -266,29 +266,52 @@ func (ss *ScaleSet) newVMSSVirtualMachinesCache(resourceGroupName, vmssName, cac
...
+		for i := len(entry.vms) - 1; i >= 0; i-- {
+			if strings.EqualFold(to.String(entry.vms[i].Name), nodeName) {
+				entry.vms = append(entry.vms[:i], entry.vms[i+1:]...)
```
Comment (feiskyer) [inline pkg/provider/azure_vmss_cache.go:286]: please acquire lock for update operations?

File: pkg/provider/azure_vmss_cache.go (hunk 1 above)
Comment (nilo19) [reply pkg/provider/azure_vmss_cache.go:283]: I don't think az.vmCache is useful when the vmType is set to vmss. But adding it here does no harm. I will add it.

File: pkg/provider/azure_vmss_cache.go (hunk 2 above)
Comment (nilo19) [reply pkg/provider/azure_vmss_cache.go:290]: done

**Review by feiskyer — APPROVED** [review]
/lgtm

**Pull request merged**